File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;
import java.io.*;

public class Main{
    public static ArrayList<ArrayList<Integer>> arr = new ArrayList<>();
    public static boolean[] visited;
    public static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());
        
        for(int i = 0; i < n+1; i++){
            arr.add(new ArrayList<>());
        }
        
        visited = new boolean[n+1];
        
        for(int i = 0; i < m; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            arr.get(a).add(b);
            arr.get(b).add(a);
        }
        
        for(int i = 1; i < n+1; i++){
            Collections.sort(arr.get(i));
        }
        
        dfs(r);
        visited = new boolean[n+1];
        sb.append("\n");
        bfs(r);
        System.out.println(String.valueOf(sb));
    }
    
    public static void dfs(int v){
        visited[v] = true;
        sb.append(v + " ");
        
        for(int i = 0; i < arr.get(v).size(); i++){
            int tmp = arr.get(v).get(i);
            if(!visited[tmp]){
                dfs(tmp);
            }
        }
    }
    
    public static void bfs(int v){
        visited[v] = true;
        Queue<Integer> queue = new LinkedList<>();
        queue.add(v);
        sb.append(v+ " ");
        
        while(!queue.isEmpty()){
            int tmp = queue.poll();
            for(int i = 0; i < arr.get(tmp).size(); i++){
                int r = arr.get(tmp).get(i);
                if(!visited[r]){
                    queue.add(r);
                    visited[r] = true;
                    sb.append(r + " ");
                }
            }
        }
    }

}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

n, m, v = map(int, sys.stdin.readline().split())
graph = [[False] * (n+1) for _ in range(n+1)]
visited1 = [False] * (n+1)
visited2 = [False] * (n+1)

# print(graph)
for i in range(m):
    a, b = map(int, sys.stdin.readline().split())
    graph[a][b] = True
    graph[b][a] = True

def dfs(v):
    visited1[v] = True
    print(v, end = " ")
    for i in range(1, n+1):
        if graph[v][i] and not visited1[i]:
            dfs(i)

def bfs(v):
    queue = deque([v])
    visited2[v] = True
    while queue:
        v = queue.popleft()
        print(v, end = " ")
        for i in range(1, n+1):
            if not visited2[i] and graph[v][i]:
                queue.append(i)
                visited2[i] = True

dfs(v)
print()
bfs(v)